MaxText/llama_or_mistral_ckpt.py

```diff
@@ -230,6 +242,174 @@ def permute_to_match_maxtext_rope(arr):
+  base_num_decoder_layers = model_params["num_layers"]
+  base_num_query_heads = model_params["num_heads"]
+  head_dim = model_params["dims_per_head"]
+  mem_info = psutil.Process()
+  logging.debug("Memory usage: %f GB", mem_info.memory_info().rss / (1024**3))
+
+  max_logging.log(f"Loading the lora  model from {lora_config['lora_model_path']}")
+  # Load LoRA model weights
+  lora_chkpt_vars = torch.load(lora_config["lora_model_path"])
+  lora_chkpt_vars = _HFNamespaceMapper(lora_chkpt_vars)
+
+  jax_weights_lora = {
+      "decoder": {
+          "layers": {
+              "mlp": {
+                  "wi_0": {
+                      "lora_a.kernel": None,
+                      "lora_b.kernel": None,
+                  },
+                  "wi_1": {
+                      "lora_a.kernel": None,
+                      "lora_b.kernel": None,
+                  },
+                  "wo": {
+                      "lora_a.kernel": None,
+                      "lora_b.kernel": None,
+                  },
+              },
+              "pre_self_attention_layer_norm": {"scale": None},
+              "post_self_attention_layer_norm": {"scale": None},
+              "self_attention": {},
+          },
+          "decoder_norm": {"scale": None},
+          "logits_dense": {"kernel": None},
+      },
+      "token_embedder": {"embedding": None},
+  }
+
+  # self attention ###############################################
+  self_attention_lora = {
+      "query": {
+          "lora_a.kernel": None,
+          "lora_b.kernel": None,
+      },
+      "key": {
+          "lora_a.kernel": None,
+          "lora_b.kernel": None,
+      },
+      "value": {
+          "lora_a.kernel": None,
+          "lora_b.kernel": None,
+      },
+      "out": {
+          "lora_a.kernel": None,
+          "lora_b.kernel": None,
+      },
+  }
+
+  lora_target_modules = lora_config["target_modules"]
+  lora_rank = int(lora_config["r"])
+  stack_shape = (base_num_decoder_layers,)
+
+  for layer_idx in range(base_num_decoder_layers):
+    for target_module in lora_target_modules:
+      if "q_proj" in target_module:
+        lora_A_q = lora_chkpt_vars[f"layers.{layer_idx}.attention.wq.lora_A.weights"].type(torch.float16).numpy().transpose()
+        lora_B_q = lora_chkpt_vars[f"layers.{layer_idx}.attention.wq.lora_B.weights"].type(torch.float16).numpy().transpose()
+
+        lora_B_q = np.reshape(lora_B_q, [lora_rank, base_num_query_heads, head_dim])
+
+        if self_attention_lora["query"]["lora_a.kernel"] is None:
+          self_attention_lora["query"]["lora_a.kernel"] = np.zeros(stack_shape + lora_A_q.shape, dtype=np.float16)
+          self_attention_lora["query"]["lora_b.kernel"] = np.zeros(stack_shape + lora_B_q.shape, dtype=np.float16)
+
+        self_attention_lora["query"]["lora_a.kernel"][layer_idx, ...] = lora_A_q  # pylint: disable=E1137
+        self_attention_lora["query"]["lora_b.kernel"][layer_idx, ...] = lora_B_q  # pylint: disable=E1137
```

Comment:
Is there any way to move this functionality to separate methods? I see they are slightly different, but there is a lot of repeated code several times in this file

Reply:
Done.